### PR TITLE
EPMRPP-33317-Defect-types-page

### PR DESCRIPTION
**Pull request by tamas-levente-szucs-epam**
screenshot 19

### CODE REVIEW

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    initialize: () => {},
+    handleSubmit: () => {},
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      defectTypes: this.props.subType,
+      isInEdit: [],
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: `is<something>` is used for boolean values

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+            Object.assign(
+              {},
+              defectType.locator === changedDefectType.locator ? changedDefectType : defectType,
+            ),
+          ),
+        }),
+    );
+  };
+
+  deleteDefectTypeRequest = (data, setState) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: setState?

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    });
+
+  showErrorNotification = (message) =>
+    this.props.showNotification({
+      message: this.props.intl.formatMessage(message),
+      type: NOTIFICATION_TYPES.ERROR,
+    });
+
+  deleteDefectType = (deletedDefectType) => () =>
+    this.deleteDefectTypeRequest(deletedDefectType, (deletedLocator) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: deleteDefectTypeRequest
deleteDefectType

Can it be a single function? Or maybe these two can have more meaningful names?

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+      .then(() => {
+        setState(data.locator);
+        this.showSuccessNotification(Messages.deleteSuccessNotification);
+      })
+      .catch(() => this.showErrorNotification(Messages.deleteErrorNotification));
+  };
+
+  createDefectTypeRequest = (data, setState) => {
+    const newData = data;
+    newData.color = this.state.newColor;
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: please don't mutate incoming arguments!

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+      .then((response) => {
+        setState({ locator: response.locator, id: response.id, ...newData });
+        this.showSuccessNotification(Messages.createSuccessNotification);
+      })
+      .catch(() => this.showErrorNotification(Messages.createErrorNotification));
+  };
+
+  changeDefectTypeRequest = (data, setState) => {
+    const newData = data;
+    newData.ids[0].color = this.state.newColor;
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: please don't mutate incoming arguments!

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+  deleteDefectTypeRequest = (data, setState) => {
+    fetch(URLS.defectTypesDelete(this.props.projectId, data.id), { method: 'delete' })
+      .then(() => {
+        setState(data.locator);
+        this.showSuccessNotification(Messages.deleteSuccessNotification);
+      })
+      .catch(() => this.showErrorNotification(Messages.deleteErrorNotification));
+  };
+
+  createDefectTypeRequest = (data, setState) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: `createDefectType`?

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    newData.color = this.state.newColor;
+    fetch(URLS.defectTypesCreate(this.props.projectId), { method: 'post', data: newData })
+      .then((response) => {
+        setState({ locator: response.locator, id: response.id, ...newData });
+        this.showSuccessNotification(Messages.createSuccessNotification);
+      })
+      .catch(() => this.showErrorNotification(Messages.createErrorNotification));
+  };
+
+  changeDefectTypeRequest = (data, setState) => {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: `changeDefectType`?

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+
+  setDefectTypeEditMode = (defectType) => (editMode) =>
+    this.setState({
+      newColor: defectType.color,
+      isInEdit: editMode
+        ? this.state.isInEdit.concat(defectType.locator)
+        : this.state.isInEdit.filter((locator) => locator !== defectType.locator),
+    });
+
+  MAX_DEFECT_SUBTYPES_SIZE = 15;
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: What'sthe difference with the `MAX_SUB_TYPES` constant?

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+      <div className={cx('button-add')}>
+        <GhostButton disabled={disabled} icon={PlusIcon} onClick={() => setEditMode(true)}>
+          {btnText}
+        </GhostButton>
+      </div>
+      <div className={cx('type-slot-left')}>{defectTypesLeftText}</div>
+    </div>
+  );
+
+  defectTypesChartData = (subTypesGroup) => ({
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: getDefectTypesChartData
calculateDefectTypesChartData

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+                    ) : (
+                      <DefectTypeInput
+                        cx={cx}
+                        form={defectType.locator}
+                        longName={defectType.longName}
+                        shortName={defectType.shortName}
+                        longNamePlaceholder={formatMessage(Messages.longNamePlaceholder)}
+                        shortNamePlaceholder={formatMessage(Messages.shortNamePlaceholder)}
+                        onFormSubmit={(editedFields) =>
+                          this.changeDefectType(group, Object.assign({}, defectType, editedFields))
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesGroup/defectTypesGroup.jsx]: `{ ...defectType, ...editedFields }`

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesMock.json [hunk 11]
```diff
@@ -0,0 +1,145 @@
+defectTypesMock = {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesMock.json]: It's not a valid json.

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+  DEFECT_TYPE_GROUP_TITLES = {
+    PRODUCT_BUG: this.props.intl.formatMessage(Messages.PRODUCT_BUG_MSG),
+    SYSTEM_ISSUE: this.props.intl.formatMessage(Messages.SYSTEM_ISSUE_MSG),
+    AUTOMATION_BUG: this.props.intl.formatMessage(Messages.AUTOMATION_BUG_MSG),
+    NO_DEFECT: this.props.intl.formatMessage(Messages.NO_DEFECT_MSG),
+    TO_INVESTIGATE: this.props.intl.formatMessage(Messages.TO_INVESTIGATE_MSG),
+  };
+
+  DefectTypesTableHeader = ({ formatMessage }) => (
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: `renderDefectTypesTableHeader`
or move it to a separate file

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 13]
```diff
@@ -0,0 +1,95 @@
+/* eslint-disable guard-for-in */
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: ???

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        {formatMessage(Messages.diagram)}
+      </div>
+    </div>
+  );
+
+  sortinSubTypes() {
+    this.subTypes = this.props.subTypes;
+    // eslint-disable-next-line guard-for-in
+    // eslint-disable-next-line no-restricted-syntax
+    for (const group in this.subTypes) {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: Object.keys

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+      this.subTypes[group] = this.subTypes[group].sort((obj1, obj2) => obj1.id - obj2.id);
+    }
+  }
+
+  render() {
+    const SUBTYPES_GROUP_SEQUENCE = [
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: is can be defined outside of `render`

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+      <div className={cx('table_column', 'table_column--color')}>
+        {formatMessage(Messages.color)}
+      </div>
+      <div className={cx('table_column', 'table_column--diagram')}>
+        {formatMessage(Messages.diagram)}
+      </div>
+    </div>
+  );
+
+  sortinSubTypes() {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: sortSubTypes

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    this.subTypes = this.props.subTypes;
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: Why this.subTypes?

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+  render() {
+    const SUBTYPES_GROUP_SEQUENCE = [
+      PRODUCT_BUG,
+      SYSTEM_ISSUE,
+      AUTOMATION_BUG,
+      NO_DEFECT,
+      TO_INVESTIGATE,
+    ];
+
+    this.sortinSubTypes();
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.jsx]: `const sortedSubTypes = this.sortSubTypes(this.props.subTypes);`

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.scss [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+  }
+  @media (max-width: $SCREEN_XS_MAX) {
+    padding: 0;
+  }
+}
+
+$table-column-padding: 11px;
+
+@mixin row-font {
+  font: bold 12px/11px $FONT-REGULAR;
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.scss]: don't set font-weight

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.scss [hunk 20]
```diff
@@ -0,0 +1,261 @@
+.settings-tab-content {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/defectTypesTab.scss]: Please split this file in different files (each component should have its' own scss file)

File: app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/modals/deleteModal/deleteModal.jsx [hunk 21]
```diff
@@ -0,0 +1,43 @@
...
+import PropTypes from 'prop-types';
+import { ModalLayout, withModal } from 'components/main/modal';
+import Parser from 'html-react-parser';
+import classNames from 'classnames/bind';
+import styles from './deleteModal.scss';
+
+const cx = classNames.bind(styles);
+
+@withModal('defectTypeDeleteModal')
+export class DeleteModal extends Component {
```
Comment (DiscoElevator) [inline app/src/pages/inside/settingsPage/settingsTabs/defectTypesTab/modals/deleteModal/deleteModal.jsx]: Can it be done using `app/src/pages/inside/common/modals/deleteItemsModal` instead?

**Comment by AmsterGet** [comment]
Closed because the component is implemented in several other PRs.

**Pull request closed without merging**